### ISSUE TITLE
Should properties be reactive too ?

### ISSUE THREAD

**Issue by hlehmann** [Urigo/angular-meteor#939]
I want to load data depending on a property of my object

with 1.2

    $scope.data = loadData() //$scope.$meteorObject(...)
    $scope.$watch('data.category', category=> loadCategoryDetails(category))

with 1.3

    $scope.helpers({
      data() {
        return loadData() //findOne ...
      }
      category() {
        return loadCategoryDetails($scope.data.category)
      }
    })

`$scope.data = {category: '1'}` will refresh category

`$scope.data.category = '1'` will not

**Comment by Deadly0** [comment]
@hlehmann looks strange. Did you tried use deep watch (send to `$watch` 3rd param `true`) or use $watchCollection instead?

**Comment by hlehmann** [comment]
well it works well using $scope.$watch but not with the helper.
The question is more : is it appropriate to use the helpers when it is combine with angular changes (like in a form)

**Comment by sebakerckhof** [comment]
+1

**Comment by Deadly0** [comment]
@hlehmann it should work with form. If you want to watch helper use `this.autorun(() => {/* your helper here */})` It's works for me.

**Comment by canac** [comment]
I am also having trouble getting changes in my scope properties to update my helpers. @Deadly0, could you give a more complete example of using autorun? Because `$scope.autorun(() => $scope.helpers({ ... }) );` is still not working for me. Is that what you meant?

**Comment by Deadly0** [comment]
@canac `selectedId` should be helper

```
const app = angular.module('test', ['angular-meteor']);

app.controller('TestCtrl', function($scope) {
  $scope.selectItem = function(item) {
    $scope.selectedId = item._id;
  };

  $scope.helpers({
    selectedId: null,
    items() {
      return Items.find();
    },

    selectedItem() {
      return Items.findOne($scope.selectedId);
    },
  });
});
```

**Comment by canac** [comment]
@Deadly0 Yes, that works, both in that contrived example and in my actual application! Thank you so much for your help!

Regarding @hlehmann's original question, I found that helpers are not deeply reactive. For example, given the following helpers:

```js
$scope.helpers({
  selectedId: { id: null },
  items() {
    return Items.find();
  },
  selectedItem() {
    return Items.findOne($scope.selectedId.id);
  },
});
```

executing `$scope.selectedId.id = 'someid'` does not cause the selectedItem to update, while executing `$scope.selectedId = { id: 'someid' }` does cause it to update. It seems that you have to change the value of the reactive helper itself, not just one of its properties to cause it to update its dependents.

**Comment by shantanubhadoria** [comment]
+1
Facing the same issue. This is a bigger problem when the value of the property in a subscription is updated using a ng-model in the template.

      this.helpers({
        partyOptions: {
          tags: []
        }
      });
      this.subscribe('parties', () => {
        return [
          this.partyOptions.tags
        ]
      });

Then in my view if I have a ng-model in a ui-select tied in to partyOptions.tags
It doesn't make the subscription trigger.

My use case is that I want one of my ui-select dropdown to reactively fill up based on another ui-select dropdown selections. but it doesn't update reactively.

**Comment by Deadly0** [comment]
@shantanubhadoria @canac have the same issue too! Quick solution is to use `$scope.getReactively('var_name', true)` but looks like it doesn't work if you use `this.subscribe` API. No ideas how to do when getReactively will be removed in 1.4

@canac you can use $scope.getReactively
```
selectedItem() {
    return Items.findOne($scope.getReactively('selectedId', true).id);
  }
```
At least it works for me in subscription.

**Comment by shantanubhadoria** [comment]
@Deadly0 
How do you mean it works for you in a subscription?
Are you not using a property of a object as a parameter in your subscription?
Thanks

**Comment by Deadly0** [comment]
@shantanubhadoria 
Here is working example where helper is an object. But, as i say before, i'm not sure will it works if you use `this.subscribe`
```
	$scope.helpers
		leaguesEnabled: {}

	$scope.subscribe 'events', () ->
		currentDateForFind = moment(Session.get('currentDate')).add(7, 'h').toDate()
		nextDate = moment(currentDateForFind).add(1, 'd').toDate()
		selectedLeagues = []
		leaguesEnabled = $scope.getReactively 'leaguesEnabled', true // <------------------

		for leagueId, enabled of leaguesEnabled
			selectedLeagues.push leagueId if enabled

		return [
			leagueId: $in: selectedLeagues
			startDate:
				$gte: currentDateForFind
				$lt: nextDate
		]
```

Also have a bug with 1.3 after `leaguesEnabled` changed it's updates the subscription, meteor loads new data, but angular do not update the view (html) until some other changes happen. Has someone faced this problem?

**Comment by canac** [comment]
In case it helps someone else, my solution was to use a `$watch`, something like this (I am sure that `$scope.getReactively` would work, but I would like to avoid using deprecated APIs as much as possible):

```js
$scope.helpers({
  selected: { id: null },
  selectedId: null,

  items() {
    return Items.find();
  },

  selectedItem() {
    return Items.findOne($scope.selectedId);
  },
});

$scope.$watch('selected.id', selectedId => {
  $scope.selectedId = selectedId;
});
```

I make a separate helper for the inner property, make my other helpers depend on that helper, and then automatically keep that helper in sync with the original object through a `$scope.$watch`. Now `$scope.selected.id = 'someid'` updates the selectedItem helper. It is not the cleanest solution, and will start to get out of hand if you need to have multiple reactive properties, but at least it works.

**Comment by canac** [comment]
@Deadly0 It sounds like you might need a call to [$scope.$apply](https://docs.angularjs.org/api/ng/type/$rootScope.Scope#$apply) to inform Angular that you have made changes to your scope properties.

**Comment by shantanubhadoria** [comment]
@Urigo Will there be a fix for this issue so we can do this elegantly in the code? I am not keen on doing intermediate hacks/workarounds in my codebase as long as the patch is imminent.
Thanks

**Comment by snowping** [comment]
I have the same problem and yes the proposed workaround works fine. I've used a different approach though. 
Using 
```
 vm.helpers({
    dummy: '',
});
$scope.$watchCollection(object, function() {
     vm.dummy =  (new Date%9e6).toString(36); //random string to tigger autorun
});  
```
I observe changes on all properties of my object and once a autorun is required I just define the dummy var to a new string to retrigger autorun. 

Of course, I would prefer a less hacky solution and since I am using controllerAs design I prefer not too have $scope variables within my controller anyways ;-)

**Comment by dotansimha** [comment]
We are thinking about providing the ability to use `ReactiveVar` as helpers. 
The second argument of `ReactiveVar` is the object equality function, just like in AngularJS `$watch`. 

This would provide the ability to create an advanced helpers like:
```js
this.helpers({
    myObject: new ReactiveVar(10, true); // the second argument could also be a method
});
```

**Comment by DmitryEfimenko** [comment]
+1. Please add support for reactive properties on the object. It would reduce the need for unnecessary hacks tremendously and make the whole thing as beautiful as a unicorn! :)

**Comment by longxuanho** [comment]
+1

**Comment by longxuanho** [comment]
With the code from @Deadly0  but when I use this instead of $scope, it is not reactive.

```javascript
const app = angular.module('test', ['angular-meteor']);

app.controller('TestCtrl', function($scope) {
  $reactive(this).attach($scope);

  this.helpers({
    selectedId: null,
    items() {
      return Items.find();
    },
    selectedItem() {
      return Items.findOne(this.selectedId);
    },
  });
});
``` 

selectedItem will not run when selectedId change. But things works well with $scope.
Now I think the $reactive not as good as I expected before.

**Comment by Urigo** [comment]
Thanks everyone for finding that out and for the solutions/workarounds.
We will release a fix for that later today to master and then release it with 1.3.1 very soon.
Sorry for the inconveniences, I hope that 1.3.1 will solve the main issues with the new version:
https://github.com/Urigo/angular-meteor/milestones/1.3.1

**Comment by shantanubhadoria** [comment]
Just to add a context on further issues this might create. I am stuck with broken reactive joins as well.

      let vm = $reactive(this).attach($scope);

      vm.helpers({
        myshelf: () => {
          return Shelfs.findOne({_id: 'abcdefgf'});
        },
        books: () => {
          return Books.find({_id: { $in: vm.myshelf.books}});
        }
      });

This too doesn't work, because we are referencing a sub object of myshelf. In addition I don't know how to add a watch over a property of myshelf. Any one found a workaround for this?
Thanks

**Comment by Urigo** [comment]
Fixed in the new [1.3.1 release](https://github.com/Urigo/angular-meteor/releases).
But as I've said in the discussion, I've added a message that you should move to use [getReactively](http://www.angular-meteor.com/api/1.3.1/get-reactively) when you want to make a regular variable reactive.
